### ISSUE TITLE
element_text size using vector

### ISSUE THREAD

**Issue by mpgerstl** [tidyverse/ggplot2#3492]
Hi,

When using a vector for axis text size in a flipped bar chart, the bars seems to be moved far outside the plot region, and also the text is shifted. I am using ggplot2 version 3.2.1. 
I was able to produce the wished plot with ggplot2 version 3.1.1

```r
library(tidyverse)
library(tibble)
ds <- mtcars %>%
  rownames_to_column() %>%
  select(car = rowname, mpg)

ds_size = rep(8, nrow(ds))
ds_size[10] <- 12

ds %>%
  ggplot(aes(x = car, y = mpg)) +
  coord_flip() +
  theme(axis.text.y = element_text(size = ds_size))
```
Thank you

**Comment by batpigandme** [comment]
Would you mind turning this into a [reprex](http://reprex.tidyverse.org/) (short for minimal **repr**oducible **ex**ample)? It's especially helpful for ggplot2, since it automatically generates and uploads the plots (or lack thereof), which makes it much easier to go through at a glance.

If you've never heard of a reprex before, you might want to start by reading the tidyverse.org [help page](https://www.tidyverse.org/help/).

Right now the best way to install reprex is:
```r
# install.packages("devtools")
devtools::install_github("tidyverse/reprex")
```
If you run into problems with access to your clipboard, you can specify an outfile for the reprex, and then copy and paste (or drag and drop) the contents here.

```r
reprex::reprex(input = "fruits_stringdist.R", outfile = "fruits_stringdist.md")
```

Thanks

**Comment by mpgerstl** [comment]
Thank you for the hint with the clipboard

``` r
library(tidyverse)
library(tibble)
ds <- mtcars %>%
  rownames_to_column() %>%
  select(car = rowname, mpg)

ds_size = rep(8, nrow(ds))
ds_size[10] <- 12

ds %>%
  ggplot(aes(x = car, y = mpg)) +
  coord_flip() +
  theme(axis.text.y = element_text(size = ds_size))
```

![](https://i.imgur.com/Z3lppHY.png)

<sup>Created on 2019-08-19 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by batpigandme** [comment]
Hmm, I get the same results you've shown above with the dev version.

Here it is with ggplot2 **3.0.0**
``` r
library(tidyverse)

ds <- mtcars %>%
  rownames_to_column() %>%
  select(car = rowname, mpg)

ds_size = rep(8, nrow(ds))
ds_size[10] <- 12

ds %>%
  ggplot(aes(x = car, y = mpg)) +
  coord_flip() +
  theme(axis.text.y = element_text(size = ds_size))
```

![](https://i.imgur.com/3Vwvmpk.png)

---------

And below is with **3.1.1**

``` r
library(tidyverse)

ds <- mtcars %>%
  rownames_to_column() %>%
  select(car = rowname, mpg)

ds_size = rep(8, nrow(ds))
ds_size[10] <- 12

ds %>%
  ggplot(aes(x = car, y = mpg)) +
  coord_flip() +
  theme(axis.text.y = element_text(size = ds_size))
```

![](https://i.imgur.com/KoVsC34.png)

<sup>Created on 2019-08-19 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by yutannihilation** [comment]
It seems the element takes the sum of the values as margin. BTW, is `element_*()` supposed to accept vectors? I've seen this kind of tricks somewhere several times, but I guess none of the official docs refer to this.

``` r
library(ggplot2)

d <- data.frame(x = letters[1:10], y = 1:10, stringsAsFactors = FALSE)

f <- function(n) {
  ggplot(d) +
    geom_point(aes(x, y)) +
    theme(axis.text.x = element_text(size = rep(8, n)))
}

f(1)
```

![](https://i.imgur.com/LCKQJyO.png)

``` r
f(10)
```

![](https://i.imgur.com/NMopmap.png)

``` r
f(100)
```

![](https://i.imgur.com/E0e0bjM.png)

<sup>Created on 2019-08-19 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by mpgerstl** [comment]
@batpigandme Sorry, I forgot the geometry in my reprex. Here is a new one

``` r
library(ggplot2)

ds <- data.frame(
  car = rownames(mtcars),
  mpg = mtcars$mpg
)

ds_size = rep(8, nrow(ds))
ds_size[10] <- 12

ggplot(ds, ggplot2::aes(x = car, y = mpg)) +
  geom_col() +
  coord_flip() +
  theme(axis.text.y = element_text(size = ds_size))
```

![](https://i.imgur.com/7ooIwMs.png)

<sup>Created on 2019-08-19 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

@yutannihilation I cannot remember where I got the hint, but using a vector for face still works

``` r
library(ggplot2)

ds <- data.frame(
  car = rownames(mtcars),
  mpg = mtcars$mpg
)

ds_face = rep("plain", nrow(ds))
ds_face[10] <- "bold"

ggplot(ds, ggplot2::aes(x = car, y = mpg)) +
  geom_col() +
  coord_flip() +
  theme(axis.text.y = element_text(face = ds_face))
```

![](https://i.imgur.com/r9GlxJt.png)

<sup>Created on 2019-08-19 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Thank you

**Comment by yutannihilation** [comment]
Yeah, it works and probably should, but I feet the individual specifications should belong to the scale, not to the theme.

@clauswilke
What do you think? I expect this kind of tricks will be unnecessary once gridtext gets on CRAN.

**Comment by clauswilke** [comment]
I'm not a big fan of setting text parameters as vectors in this way, mostly because it's unpredictable when it works and when it doesn't. It works for axes but it doesn't work for legends, for internal technical reasons that are not visible to the enduser.

Having said that, I don't understand why this specific example fails. We should track down the cause, because it suggests something weird is going on with the `size` argument.

**Comment by yutannihilation** [comment]
Thanks, agreed. Let's see what's happening here.

**Comment by clauswilke** [comment]
Here is what I *think* is happening. (I don't quite understand yet why it only happens for font size, and not for font face, so maybe I'm wrong.)

This line returns now a vector of descent values rather than just one value:
https://github.com/tidyverse/ggplot2/blob/1ae034ee6f5c669f497c96d3258cf2ca13907953/R/margins.R#L69
``` r
ggplot2:::font_descent("", "plain", rep(10, 5), 1)
#> [1] 0.0292290581597222inches 0.0292290581597222inches
#> [3] 0.0292290581597222inches 0.0292290581597222inches
#> [5] 0.0292290581597222inches
```

<sup>Created on 2019-08-19 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

This vector results in vectors `text_width` and `text_height`, which are given to `add_margins()` here:
https://github.com/tidyverse/ggplot2/blob/1ae034ee6f5c669f497c96d3258cf2ca13907953/R/margins.R#L197-L205

And the `add_margins()` function then build layouts assuming `width` and `height` are single values rather than vectors:
https://github.com/tidyverse/ggplot2/blob/1ae034ee6f5c669f497c96d3258cf2ca13907953/R/margins.R#L115-L121

Again, this doesn't yet fully explain what exactly is happening, but I think it's on the right track. Importantly, it's based on a line that was changed with the latest release:
https://github.com/tidyverse/ggplot2/commit/f5a88a71e7b60b6531ef6c4a936370b5e3c4ef92#diff-004d6436e77ba3e9705ab23890dace32

The previous function `descentDetails()` produced the wrong results, but it didn't create vectors.

**Comment by yutannihilation** [comment]
Thanks. For facet strips, it seems we do the right thing to pass a single `height` and `width` to `add_margin()` even when `height` and/or `width` is a vector.

https://github.com/tidyverse/ggplot2/blob/b8420241309c8eea00d7086002c01cdf38a50eac/R/labeller.r#L608-L631

**Comment by clauswilke** [comment]
I made a PR that seems to fix the problem, though I still don't fully understand how exactly things go wrong. #3493

**Comment by yutannihilation** [comment]
Confirmed #3493 fixes the issue. But, I too don't understand how the things are supposed to work here...

Is `font_descent()` really intended to work with multiple `size`s? I see it returns the same values for different sizes in the vector (Does this happen only on Windows?).

``` r
ggplot2:::font_descent("", "plain", 10, 1)
#> [1] 0.03125inches
ggplot2:::font_descent("", "plain", c(10, 10), 1)
#> [1] 0.03125inches 0.03125inches

ggplot2:::font_descent("", "plain", c(10, 20), 1)
#> [1] 0.03125inches 0.03125inches
ggplot2:::font_descent("", "plain", 1:10 * 10, 1)
#>  [1] 0.03125inches 0.03125inches 0.03125inches 0.03125inches 0.03125inches
#>  [6] 0.03125inches 0.03125inches 0.03125inches 0.03125inches 0.03125inches

# probably the cache works wrong here...
ggplot2:::font_descent("", "plain", 1:10 * 20, 1)
#>  [1] 0.03125inches            0.03125inches           
#>  [3] 0.03125inches            0.03125inches           
#>  [5] 0.03125inches            0.0520833333333333inches
#>  [7] 0.0520833333333333inches 0.0520833333333333inches
#>  [9] 0.0520833333333333inches 0.0520833333333333inches
```

<sup>Created on 2019-08-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by clauswilke** [comment]
Yes, the font cache is not working correctly for vectorized input. It loops over the keys but then uses the same parameters for descent lookup each time:
https://github.com/tidyverse/ggplot2/blob/b8420241309c8eea00d7086002c01cdf38a50eac/R/margins.R#L339-L355

@thomasp85 Do you have any strong opinions on how to fix this? Since `font_descent()` is only used by `title_spec()` and that is not vectorized, maybe it's best to let `font_descent()` return only the first result? In general, the fact that `title_spec()` is not properly vectorized causes all sorts of problems, but fixing that is a pretty substantial rewrite, and probably better left to something like the gridtext package.

**Comment by clauswilke** [comment]
I made a new PR that simplifies `font_descent()` by only returning the first result. I'm Ok with that, because even if it returned multiple results correctly the `title_spec()` function wouldn't know what to do with them.

**Comment by yutannihilation** [comment]
(Just a note to self)

> It seems the element takes the sum of the values as margin.

This behaviour comes from here; `widthDetails()` and `heightDetails()` takes the sum of the widths and heights of a `titleGrob`.

https://github.com/tidyverse/ggplot2/blob/b8420241309c8eea00d7086002c01cdf38a50eac/R/margins.R#L214-L216

**Comment by clauswilke** [comment]
But why do we see the behavior only for font size but not for example for font face? I still don't get that.

**Comment by yutannihilation** [comment]
Curiously, it seems another bug... Here, `gp$fontface` is passed to `title_spec()`, but `gp` contains `font`, not `fontface`.

https://github.com/tidyverse/ggplot2/blob/b8420241309c8eea00d7086002c01cdf38a50eac/R/margins.R#L69

e.g.

``` r
library(ggplot2)

d <- data.frame(x = letters[1:10], y = 1:10, stringsAsFactors = FALSE)

ggplot(d) +
  geom_point(aes(x, y)) +
  theme(axis.text.x = element_text(size = 30, face = rep("plain", 10)))
```

![](https://i.imgur.com/CjbHeTy.png)

``` r

ggplot(d) +
  geom_point(aes(x, y)) +
  theme(axis.text.x = element_text(size = 30, family = rep("", 10)))
```

![](https://i.imgur.com/WKj5YqU.png)

<sup>Created on 2019-08-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by yutannihilation** [comment]
It seems `gpar()` treats `fontface` differently than other parameters.

``` r
grid::gpar(
  fontsize = NULL,
  col = NULL,
  fontfamily = NULL,
  fontface = NULL,
  lineheight = NULL
)
#> $fontface
#> NULL
```

<sup>Created on 2019-08-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

As `gp` keeps `fontface = NULL`, it removes `fontface` from `element_gp` at `modify_list(element_gp, gp)` here:

https://github.com/tidyverse/ggplot2/blob/b8420241309c8eea00d7086002c01cdf38a50eac/R/theme-elements.r#L209-L219

**Comment by clauswilke** [comment]
Ok, I see it now. The strange handling of `fontface` in `gpar()` is actually one of my number one pet peeves. It takes so much effort to work around it. I'm constantly battling with it in gridtext.

**Comment by mpgerstl** [comment]
I have tested the different parameters for `element_text()` in the reprex above using vectors.
It is working for: `face`, `colour`, `hjust` and `vjust` - whereas the font int hjust did not render well
it is not working for: `size`, `angle` and `family`

**Comment by yutannihilation** [comment]
> it is not working for: `size`, `angle` and `family`

Opps, thanks...

@clauswilke 
#3493 doesn't fix `angle`. We need to consider the case when `angle` is a vector here:

https://github.com/tidyverse/ggplot2/blob/b8420241309c8eea00d7086002c01cdf38a50eac/R/margins.R#L74-L75

**Comment by clauswilke** [comment]
Ok. I think we should just ignore all but the first angle. I expect using multiple angles along the axis never worked (but I've never tried it).

**Comment by yutannihilation** [comment]
Agreed.

> I expect using multiple angles along the axis never worked (but I've never tried it).

I confirmed it didn't work at the time of v3.1.1.

**Comment by thomasp85** [comment]
A bit late to this. I honestly think we shouldn’t allow vectorised input in `element_text()`. As mentioned the behaviour is unpredictable but when it works by chance it gives the impression that it is a feature.

**Comment by thomasp85** [comment]
There are valid obscure reason for the weirdness of the `fontface` element in `gpar` and that is not going to change. For performance reasons the gpar list is treated specially in the grid C internals and fiddling around with it can mess things up unexpectedly

**Comment by yutannihilation** [comment]
Do you remember how many revdep packages broke when you tried to stop support vectorised input? https://github.com/tidyverse/ggplot2/pull/3237

**Comment by thomasp85** [comment]
Sounds like it was just a single package but I honestly can’t remember. I’d propose that we catch this at the `element_*()` and throw a warning so that it is clear what is not allowed/supported. We will never support this behaviour fully anyway

**Comment by yutannihilation** [comment]
OK, if it's just one or two packages, I agree with the idea of warning. It sounds the right approach.

**Comment by clauswilke** [comment]
Let's not say "never". My `element_markdown()` from ggtext calculates margins, descents, grob extents, etc. correctly for vectorized input (not that the example necessarily shows that).

``` r
library(ggplot2)
library(ggtext)

d <- data.frame(x = letters[1:16], y = 1:16, stringsAsFactors = FALSE)

ggplot(d) +
  geom_point(aes(x, y)) +
  theme(axis.text.x = element_markdown(size = 3*c(5:12, 11:4)))
```

![](https://i.imgur.com/CoGFTHm.png)

<sup>Created on 2019-08-24 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Also, I feel that the whole `title_spec()` code is a bit of a mess and should be replaced at some point.

I'll add a warning at the drawing stage, so we get a warning exactly where the code receives input it can't handle. This way, if at some point in the future we have more general support for vectorized input, the warning will simply disappear.

**Comment by thomasp85** [comment]
I can only agree with you on `title_spec()`. What I mean with *never* is that while the `element_*()` might work vectorised it will depend on the context in which they are called. Your `element_markdown()` will not behave as if it was vectorised for strip texts so having it work that way for axis text will only confuse. 

On a deeper level, I really don’t like that the theming can become tightly coupled with the exact plot, eg your example will only work with plots that have 16 x-axis labels. I don’t think that should be allowed

**Comment by hadley** [comment]
I think @thomasp85’s reasoning is compelling: a theme should not be tied to a specific plot, and any use of vectors should be considered an unsupported hack. 

That said, I think people we be unhappy if we remove this without warning so I think it’ll have to go through a deprecation cycle.

**Comment by thomasp85** [comment]
I agree we should deprecate it carefully, but we should at the least not build out support for the hack🙂

**Comment by thomasp85** [comment]
FWIW, I think most of the use cases for this hack can be solved properly using `element_markdown()`

**Comment by clauswilke** [comment]
I've updated my PR. How is this?

``` r
library(ggplot2)

d <- data.frame(x = letters[1:16], y = 1:16, stringsAsFactors = FALSE)

ggplot(d) +
  geom_point(aes(x, y)) +
  theme(axis.text.x = element_text(size = 3*c(5:12, 11:4)))
#> Warning: Vectorized input to `element_text()` is not officially supported.
#> Results may be unexpected or may change in future versions of ggplot2.
```

![](https://i.imgur.com/yUlei8k.png)

<sup>Created on 2019-08-25 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by yutannihilation** [comment]
> `#> Results may be unexpected or may change in future versions of ggplot2.`

If we agree to put the vectorized behavior into the deprecation cycle, I think we should a bit stronger words to discourage the usage.

**Comment by yutannihilation** [comment]
(Just a side comment)

> On a deeper level, I really don’t like that the theming can become tightly coupled with the exact plot

I too felt so, but, after some thinking, now I'm not sure if I can fully agree with this. Leaving the technical details, a theme *can* be tightly coupled with the exact plot; For example, it depends on the length of labels how much angle we should rotate the labels. Using the analogy of HTML and CSS, they are designed to separate theming from the data, but, in practice, elaborated CSSs are so tightly coupled with the exact HTMLs that they cannot be reused easily.

**Comment by thomasp85** [comment]
I don’t think your example is exactly fitting. What I’m talking about is a deep “technical” coupling — not an aesthetic one. True, label rotation may only look good with packed labels, but there are no technical requirements of the plot that means that a plot will fail to render with certain rotation settings.

**Comment by yutannihilation** [comment]
> there are no technical requirements of the plot that means that a plot will fail to render with certain rotation settings.

Ah, I see.
